lib/spack/spack/solver/asp.py

```diff
@@ -3723,6 +3832,102 @@ def _order_index(flag_group):
     def deprecated(self, node: NodeArgument, version: str) -> None:
         tty.warn(f'using "{node.pkg}@{version}" which is a deprecated version')
 
+    def splice_at_hash(
+        self,
+        parent_node: NodeArgument,
+        splice_node: NodeArgument,
+        child_name: str,
+        child_hash: str,
+    ):
+        splice = Splice(splice_node, child_name=child_name, child_hash=child_hash)
+        if parent_node not in self._splices:
+            self._splices[parent_node] = []
+        self._splices[parent_node].append(splice)
+
+    def splice_in_dependency(
+        self, parent_node: NodeArgument, child_node: Union[str, NodeArgument]
+    ):
+        if parent_node not in self._splice_dag:
+            self._splice_dag[parent_node] = []
+        # Hash is the base-case covered by splice_at_hash
+        if isinstance(child_node, NodeArgument):
+            self._splice_dag[parent_node].append(child_node)
+
+    def _resolve_splices_for_node(
+        self, node: NodeArgument, resolved: Dict[NodeArgument, spack.spec.Spec]
+    ) -> spack.spec.Spec:
+        """
+        This function both caches it's answer in resolved, and returns the
+        spec resulting from resolving splices in the nodes
+        """
+        # Bottom up caching
+        if node in resolved:
+            return resolved[node]
+        orig_spec = self._specs[node]
+        immediate_splices = self._splices.get(node, [])
+        deps_with_splices = self._splice_dag.get(node, [])
+        # This node has no splicing to be done to it.
+        if len(immediate_splices) == 0 and len(deps_with_splices) == 0:
```

Comment:
```suggestion
        if not immediate_splices and not deps_with_splices:
```
slightly clearer and more pythonic